File: C++/level1/크레인_인형뽑기_게임.cpp
#include <string>
#include <vector>
#include <stack>

using namespace std;

int solution(vector<vector<int>> board, vector<int> moves) {
    int answer = 0;
    stack<int> s;
    vector<stack<int>> remain;

    for(int i=0; i<board[0].size(); i++){
        stack<int> temp;
        for(int j=board.size()-1; j>=0; j--){
            if(board[j][i] != 0)
                temp.push(board[j][i]);
            else
                break;
        }
        remain.push_back(temp);
    }

    for(int i=0; i<moves.size(); i++){
        if(!((s.empty() && remain[moves[i]-1].empty()) || remain[moves[i]-1].empty()))
            if(s.empty()){
                s.push(remain[moves[i]-1].top());
                remain[moves[i]-1].pop();
            }
            else if(remain[moves[i]-1].top() == s.top()){
                s.pop();
                remain[moves[i]-1].pop();
                answer += 2;
            }
            else{
                s.push(remain[moves[i]-1].top());
                remain[moves[i]-1].pop();
            }
    }

    return answer;
}

File: JavaScript/level1/크레인_인형뽑기_게임.js
function solution(board, moves) {
    var answer = 0;
    let stack = [];
    
    for(let i=0; i<moves.length; i++){
        for(let j=0; j<board[0].length; j++){
            if(board[j][moves[i]-1] !== 0){
                if(stack[stack.length-1] !== board[j][moves[i]-1])
                    stack.push(board[j][moves[i]-1]);
                else{
                    stack.pop();
                    answer += 2;
                }
                board[j][moves[i]-1] = 0;
                break;
            }
        }
    }
    
    return answer;
} 
